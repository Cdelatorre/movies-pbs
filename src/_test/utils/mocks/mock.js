export const moviesData = {
  viewedMovies: [
    {
      id: "1",
      title: "Movie 1",
      genres: ["Horror", "Comedy"],
      viewed: true,
      img: "https://picsum.photos/id/118/300/400",
      createdAt: "2022-10-16T20:25:58.420Z",
    },
  ],
  unViewedMovies: [
    {
      id: "2",
      title: "Movie 2",
      genres: ["Horror", "Comedy"],
      viewed: false,
      img: "https://picsum.photos/id/159/300/400",
      createdAt: "2022-10-16T20:51:29.081Z",
    },
    {
      id: "3",
      title: "Movie 3",
      genres: ["Romance"],
      viewed: false,
      img: "https://picsum.photos/id/122/300/400",
      createdAt: "2022-10-16T20:51:45.249Z",
    },
    {
      id: "4",
      title: "Movie 4",
      genres: ["Test"],
      viewed: false,
      img: "https://picsum.photos/id/168/300/400",
      createdAt: "2022-10-16T20:52:14.784Z",
    },
    {
      id: "5",
      title: "Movie 5",
      genres: ["Random"],
      viewed: false,
      img: "https://picsum.photos/id/149/300/400",
      createdAt: "2022-10-16T20:52:36.529Z",
    },
  ],
};

export const singleMovieData = {
  id: "1",
  title: "Movie 1",
  genres: ["Horror", "Comedy", "Romance"],
  viewed: false,
  img: "https://picsum.photos/id/149/300/400",
  createdAt: "2022-10-16T20:52:36.529Z",
};

export const activeFiltersData = {
  activeFilters: [],
  search: "",
};

export const activeLoadingData = {
  value: false,
};

export const emptyStoreData = {
  movies: { viewedMovies: [], unViewedMovies: [] },
  filters: { activeFilters: [], search: "" },
  loading: { value: false },
};

export const sampleFiltersData = ["Horror", "Comedy", "Romance"];

export const customStoreData = (movies, activeFilters, loading) => {
  return {
    movies: movies ? moviesData : emptyStoreData.movies,
    filters: activeFilters ? activeFiltersData : emptyStoreData.filters,
    loading: loading ? activeLoadingData : emptyStoreData.loading,
  };
};
