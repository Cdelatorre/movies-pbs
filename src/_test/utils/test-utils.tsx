import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../../store";
import { BrowserRouter, Router } from "react-router-dom";
import { emptyStoreData } from "./mocks/mock";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = emptyStoreData,

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route?: string
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (route) window.history.pushState({}, "Test routing", route);

    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>;
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouter(ui: React.ReactElement, history?: any) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (history) {
      return (
        <Router location="/" navigator={history}>
          {children}
        </Router>
      );
    } else {
      return <BrowserRouter>{children}</BrowserRouter>;
    }
  }
  return { ...render(ui, { wrapper: Wrapper }) };
}
