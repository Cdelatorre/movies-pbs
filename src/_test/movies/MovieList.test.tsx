import { screen } from "@testing-library/react";
import MovieList from "../../components/movies/list";
import { customStoreData, moviesData } from "../utils/mocks/mock";
import { renderWithProviders } from "../utils/test-utils";

describe("Renders MovieList component correctly", () => {
  test("Renders default MovieList and each component correctly", () => {
    renderWithProviders(<MovieList movies={[]} />);
    const movieList = screen.getByTestId("movie-list");

    expect(movieList).toBeInTheDocument();
    expect(
      screen.getByRole("heading", {
        name: "Vaya, no hemos encontrado películas...",
      })
    ).toBeInTheDocument();

    expect(movieList).toMatchSnapshot();
  });

  test("Renders MovieCard component for each movie passed as prop", () => {
    const movies = [...moviesData.unViewedMovies, ...moviesData.viewedMovies];
    renderWithProviders(<MovieList movies={movies} />);

    expect(screen.getByTestId("movie-list")).toBeInTheDocument();
    expect(
      screen.queryByRole("heading", {
        name: "Vaya, no hemos encontrado películas...",
      })
    ).toBeNull();

    movies.forEach((movie: Movie) => {
      expect(screen.getByTestId(`movie-card-${movie.id}`)).toBeInTheDocument();
    });
  });

  test("Renders Movie Skeleton component for each movie passed as prop if loading is true", () => {
    const movies = [...moviesData.unViewedMovies, ...moviesData.viewedMovies];
    const customStore = customStoreData(true, false, true);
    renderWithProviders(<MovieList movies={movies} />, {
      preloadedState: {
        ...customStore,
      },
    });

    expect(screen.getByTestId("movie-list")).toBeInTheDocument();

    movies.forEach((movie: Movie) => {
      expect(
        screen.getByTestId(`movie-skeleton-${movie.id}`)
      ).toBeInTheDocument();
    });
  });
});
