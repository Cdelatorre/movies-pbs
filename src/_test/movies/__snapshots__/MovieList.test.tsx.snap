// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders MovieList component correctly Renders default MovieList and each component correctly 1`] = `
<div
  class="row mb-5"
  data-testid="movie-list"
>
  <div
    class="container"
  >
    <h4
      class="text-light mb-5 p-5 border rounded border-light"
    >
      Vaya, no hemos encontrado pel√≠culas...
    </h4>
  </div>
</div>
`;
